guidelines/general/branch_structure.md

```diff
@@ -0,0 +1,69 @@
+# Recommendations for project branch structure
+
+We have three basic branch structures that we employ, according to the requirements of each project. They are, in order of increasing complexity and overhead:
+
+1. One-branch projects
+2. "master" + "stable" branch projects
```

Comment:
Yeah, it's possible we might be able to automate that, although *if* we've applied hotfixes to "stable" *and* they end up causing merge conflicts then we'll still have to do manual resolution.

This is what I've been doing in projects with "master" + "stable" branches:

1. If "master" is in a good state (ie. ready to release, CI green), run `yarn version` (this creates the final release commit and tags it).
2. `git push $REMOTE master` directly 😱 — safe only because master was green on the previous commit, and the commit I am pushing only changes the version number and nothing else.
3. Create a *draft* PR (not intended to be merged) asking to merge "master" into "stable"; the sole purpose of this is to see CI green one last time before pushing the tag.
4. Once CI is green, `git checkout stable; git merge master` (usually with `--ff-only` because it almost always should be); then `git push $REMOTE stable --follow-tags`.
5. Finally, `npm publish` (or `npm publish --dry-run` if I feel it's necessary to double-check).

The thing about this flow is it means releasing doesn't produce a merge commit, so it produces the "best" result in the repo. The downside is that you have to be careful not to screw up, which is why we've documented it carefully in CONTRIBUTING.md files like [this one](https://github.com/liferay/liferay-js-themes-toolkit/blob/ea2e3fe13a7e6a1d6f68803e091acadbab20a9e4/CONTRIBUTING.md) (liferay-js-themes-toolkit) or [this one](https://github.com/liferay/liferay-npm-tools/blob/25733d82dbab8b1278743d653799a2682c832359/CONTRIBUTING.md) (liferay-npm-tools).

One of the reasons it is preferable to avoid merge commits for each release is that don't have to worry about explicitly excluding them in our [automated changelog generator](https://github.com/liferay/liferay-js-themes-toolkit/blob/ea2e3fe13a7e6a1d6f68803e091acadbab20a9e4/scripts/changelog.js) (there's [an issue](https://github.com/liferay/liferay-js-themes-toolkit/issues/329) for adding special-casing to handle that anyway, but still, it's nice to have a clean history). (And as far as changelogs go, I will of course move that changelog generator somewhere it can be easily re-used. It handles repos with multiple branches nicely, and if we use good PR titles, we get a good results from it — examples: [alloy-editor](https://github.com/liferay/alloy-editor/releases), [liferay-js-themes-toolkit](https://github.com/liferay/liferay-js-themes-toolkit/releases)).

Reply:
Hm, this is a big step 😅.